File: backend/spring-boot/src/main/java/org/bootstrapbugz/api/auth/security/user/details/UserPrincipal.java
package org.bootstrapbugz.api.auth.security.user.details;

import com.fasterxml.jackson.annotation.JsonIgnore;
import java.io.Serial;
import java.time.LocalDateTime;
import java.util.Collection;
import java.util.List;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import org.bootstrapbugz.api.user.model.User;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

@Getter
@AllArgsConstructor
@EqualsAndHashCode
public class UserPrincipal implements UserDetails {
  @Serial private static final long serialVersionUID = 5954870422841373076L;
  private final Long id;
  private final String firstName;
  private final String lastName;
  private final String username;
  @JsonIgnore private final String email;
  @JsonIgnore private final String password;
  private final boolean enabled;
  private final boolean accountNonLocked;
  private final LocalDateTime createdAt;
  private final Collection<? extends GrantedAuthority> authorities;

  public static UserPrincipal create(User user) {
    return new UserPrincipal(
        user.getId(),
        user.getFirstName(),
        user.getLastName(),
        user.getUsername(),
        user.getEmail(),
        user.getPassword(),
        user.getActive(),
        !user.getLock(),
        user.getCreatedAt(),
        getAuthorities(user));
  }

  private static List<GrantedAuthority> getAuthorities(User user) {
    return user.getRoles().stream()
        .map(role -> (GrantedAuthority) new SimpleGrantedAuthority(role.getName().name()))
        .toList();
  }

  @Override
  public boolean isAccountNonExpired() {
    return true;
  }

  @Override
  public boolean isCredentialsNonExpired() {
    return true;
  }
}

File: backend/spring-boot-kotlin/src/main/java/org/bootstrapbugz/api/auth/security/UserPrincipal.kt
package org.bootstrapbugz.api.auth.security

import org.bootstrapbugz.api.user.model.User
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import java.time.LocalDateTime

data class UserPrincipal(
    val id: Long? = null,
    private val firstName: String?,
    private val lastName: String?,
    private val username: String?,
    private val email: String?,
    private val password: String?,
    private val enabled: Boolean,
    private val accountNonLocked: Boolean,
    private val createdAt: LocalDateTime?,
    private val authorities: Collection<GrantedAuthority>?
) : UserDetails {

    companion object {
        private const val serialVersionUID = 5954870422841373076L

        fun create(user: User): UserPrincipal {
            val authorities = user.roles?.map { SimpleGrantedAuthority(it.name.name) }
            return UserPrincipal(
                user.id,
                user.firstName,
                user.lastName,
                user.username,
                user.email,
                user.password,
                user.active,
                !user.lock,
                user.createdAt,
                authorities
            )
        }
    }

    override fun getUsername(): String? = username
    override fun getPassword(): String? = password
    override fun getAuthorities(): Collection<GrantedAuthority>? = authorities
    override fun isEnabled(): Boolean = enabled
    override fun isAccountNonExpired(): Boolean = true
    override fun isAccountNonLocked(): Boolean = accountNonLocked
    override fun isCredentialsNonExpired(): Boolean = true
}